### ISSUE TITLE
When deploying karpenter, doing a "destroy" does not remove the EC2 instances.

### ISSUE THREAD

**Issue by veggiemonk** [volvo-cars/lingon#97]
Since those instances are not the state, terraform doesn't remove them and the EKS cluster cannot be fully destroyed.

**Comment by jlarfors** [comment]
Karpenter does set finzliers on the nodes that it creates. Docs: https://karpenter.sh/preview/troubleshooting/#uninstallation

Karpenter in our case was running on EKS Fargate nodes. Either the "destroy" of a Fargate profile (and hence nodes) or the uninstall of Karpenter is not cleaning these up.

One way to test is to uninstall Karpenter before doing a "terraform destroy" and double check that Karpenter does clean up the nodes. If it does, we know the issue lies with the destruction of the fargate profile...

**Comment by veggiemonk** [comment]
could this be related? https://aws.github.io/aws-eks-best-practices/karpenter/#amazon-eks-private-cluster-without-outbound-internet-access

We haven't allowed outbound access.

**Comment by jlarfors** [comment]
> could this be related? https://aws.github.io/aws-eks-best-practices/karpenter/#amazon-eks-private-cluster-without-outbound-internet-access
> 
> We haven't allowed outbound access.

All the clusters we created so far have been public, so I don't think so.

**Comment by jlarfors** [comment]
This seems very relevant and describes the issue: https://github.com/aws/karpenter/issues/1955

TL;DR, if we delete the provisioners as a separate cleanup step, Karpenter will delete the nodes it manages. If we just delete the cluster (without an explicit provisioner deletion step), it is likely that Karpenter will not be able to delete the nodes it manages...
